File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/plugin/KotlinRulesDefinition.java
/*
 * SonarSource SLang
 * Copyright (C) 2018-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonarsource.kotlin.plugin;

import java.util.ArrayList;
import org.sonar.api.server.rule.RulesDefinition;
import org.sonarsource.analyzer.commons.RuleMetadataLoader;
import org.sonarsource.slang.checks.CommentedCodeCheck;
import org.sonarsource.slang.checks.utils.Language;
import org.sonarsource.slang.plugin.RulesDefinitionUtils;

public class KotlinRulesDefinition implements RulesDefinition {

  private static final String RESOURCE_FOLDER = "org/sonar/l10n/kotlin/rules/kotlin";

  @Override
  public void define(Context context) {
    NewRepository repository = context
      .createRepository(KotlinPlugin.KOTLIN_REPOSITORY_KEY, KotlinPlugin.KOTLIN_LANGUAGE_KEY)
      .setName(KotlinPlugin.REPOSITORY_NAME);
    RuleMetadataLoader ruleMetadataLoader = new RuleMetadataLoader(RESOURCE_FOLDER, KotlinProfileDefinition.PATH_TO_JSON);

    ArrayList<Class<?>> checks = new ArrayList<>(KotlinCheckList.checks());
    checks.add(CommentedCodeCheck.class);
    ruleMetadataLoader.addRulesByAnnotatedClass(repository, checks);

    RulesDefinitionUtils.setDefaultValuesForParameters(repository, checks, Language.KOTLIN);

    repository.done();
  }

}

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/plugin/KotlinRulesDefinition.kt
/*
 * SonarSource Kotlin
 * Copyright (C) 2018-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
package org.sonarsource.kotlin.plugin

import org.sonar.api.SonarRuntime
import org.sonar.api.server.rule.RulesDefinition
import com.sonarsource.plugins.kotlin.api.KotlinPluginExtensionsProvider
import org.sonarsource.kotlin.api.common.KOTLIN_LANGUAGE_KEY
import org.sonarsource.kotlin.api.common.KOTLIN_REPOSITORY_KEY
import org.sonarsource.kotlin.gradle.KOTLIN_GRADLE_CHECKS

class KotlinRulesDefinition(
    private val runtime: SonarRuntime,
    private val extensionsProviders: Array<KotlinPluginExtensionsProvider>,
) : RulesDefinition {

    companion object {
        private const val RESOURCE_FOLDER = "org/sonar/l10n/kotlin/rules/kotlin"
    }

    override fun define(context: RulesDefinition.Context) {
        val checks: MutableList<Class<*>> = mutableListOf()
        checks.addAll(KOTLIN_CHECKS)
        checks.addAll(KOTLIN_GRADLE_CHECKS)

        val extensions = KotlinPluginExtensions(extensionsProviders)

        val rulesByRepositoryKey: MutableMap<String, List<Class<*>>> = mutableMapOf()
        rulesByRepositoryKey.putAll(extensions.rulesByRepositoryKey())
        rulesByRepositoryKey[KOTLIN_REPOSITORY_KEY] = checks

        val ruleMetadataLoaderClass = extensions.ruleMetadataLoaderClass()
        val ruleMetadataLoader = ruleMetadataLoaderClass
            .getConstructor(String::class.java, String::class.java, SonarRuntime::class.java)
            .newInstance(RESOURCE_FOLDER, KotlinProfileDefinition.PATH_TO_JSON, runtime)
        val registerRules = { repositoryKey: String, repositoryName: String ->
            val repository = context.createRepository(repositoryKey, KOTLIN_LANGUAGE_KEY)
                .setName(repositoryName)
            ruleMetadataLoaderClass
                .getMethod("addRulesByAnnotatedClass", RulesDefinition.NewRepository::class.java, List::class.java)
                .invoke(ruleMetadataLoader, repository, rulesByRepositoryKey[repositoryKey])
            repository.done()
        }
        registerRules(KOTLIN_REPOSITORY_KEY, KotlinPlugin.REPOSITORY_NAME)
        extensions.repositories().forEach(registerRules)
    }
}
